module/state_synchronization/execution_data_cid_cache.go

```diff
@@ -0,0 +1,109 @@
+package state_synchronization
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type ExecutionDataCIDCache interface {
+	Get(c cid.Cid) (BlobRecord, error)
+	Insert(header *flow.Header, blobTree BlobTree)
+}
+
+type ExecutionDataCIDCacheImpl struct {
+	mu            sync.RWMutex
+	blobs         map[cid.Cid]BlobRecord
+	blobTrees     []*BlobTreeRecord
+	blobTreeIndex int
+	size          uint
+}
+
+func NewExecutionDataCIDCache(size uint) *ExecutionDataCIDCacheImpl {
+	return &ExecutionDataCIDCacheImpl{
+		blobs:     make(map[cid.Cid]BlobRecord),
+		blobTrees: make([]*BlobTreeRecord, 0, size),
+		size:      size,
+	}
+}
+
+var ErrCacheMiss = errors.New("CID not found in cache")
+
+type BlobTreeRecord struct {
+	BlockID     flow.Identifier
+	BlockHeight uint64
+	BlobTree    BlobTree
+}
+
+type BlobTreeLocation struct {
+	Height uint
+	Index  uint
+}
+
+type BlobRecord struct {
+	BlobTreeRecord   *BlobTreeRecord
+	BlobTreeLocation BlobTreeLocation
+}
+
+func (e *ExecutionDataCIDCacheImpl) Get(c cid.Cid) (BlobRecord, error) {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+
+	if record, ok := e.blobs[c]; ok {
+		return record, nil
+	} else {
+		return BlobRecord{}, ErrCacheMiss
+	}
+}
+
+func (e *ExecutionDataCIDCacheImpl) insertBlobTree(header *flow.Header, blobTree BlobTree) *BlobTreeRecord {
+	blobTreeRecord := &BlobTreeRecord{
+		BlockID:     header.ID(),
+		BlockHeight: header.Height,
+		BlobTree:    blobTree,
+	}
+
+	if len(e.blobTrees) < int(e.size) {
+		e.blobTrees = append(e.blobTrees, blobTreeRecord)
+	} else {
+		evictee := e.blobTrees[e.blobTreeIndex]
+
+		for _, cids := range evictee.BlobTree {
```

Comment:
this seems like a lot of work to evict one item

Reply:
each CID is stored as a separate entry, and when we evict an entire blob tree we need to evict all CID's in the tree